content/ytt/docs/latest/yaml-primer.md

```diff
@@ -0,0 +1,195 @@
+
+An "array" is a list of "array items" (zero-indexed).
+
+Like documents (and map items), an array item has exactly one (1) value. Once more, it's either:
+- a **Map**
+- an **Array**, or
+- a **Scalar**
+
+![](/images/ytt/yaml-primer/map-item-has-array-ast.jpg)
+<!-- source: https://miro.com/app/board/o9J_lIfcKZY=/?moveToWidget=3074457358161105960&cot=14 -->
+_Figure 7: The **map item** "foo" has an **array** (dotted gold) for a value; that array has three **array items** (solid gold)._
+
+Let's see what it looks like when an **array item**'s value is a **map**...
+
+![](/images/ytt/yaml-primer/array-item-has-map-ast.jpg)
+<!-- source: https://miro.com/app/board/o9J_lIfcKZY=/?moveToWidget=3074457358161105960&cot=14 -->
+_Figure 8: "foo" has an **array** with one **array item** whose value is a **map** containing two **map item**s: `name` and `factor`._
+
+We've seen scalars in use above. Let's cover them explicitly.
+
+### Scalars
+
+A "scalar" is a fundamental value. YAML supports:
+- integers (e.g. 13, 42, 137, 32767)
+- floating point numbers (e.g. 1.0, 3.14, 137.03599913)
+- strings (e.g. "", "fine structure", "$ecr3t")
+- boolean values: `true` or `false`
+- `null` (when a value is omitted, that implies `null`; `~` == `null`)
+
+### Summary
+
+In short:
+
+- a YAML file is parsed into a **Document Set** which is merely a list of **Document**s.
+- a given **Document** has one _value_.
+- Both **Map**s and **Array**s are nothing more than collections of items.
+  - a **Map Item** has a _key_ and a _value_
+  - an **Array Item** has a _value_
+- a _value_ (whether held by a document, map item, or array item) is either a **Map**, an **Array**, or a **Scalar**.
+
+With this structure in mind, we can now look into how `ytt` annotates it.
+
+## Annotating YAML
+
+A `ytt` "annotation" is a YAML comment of a specific format; it attaches behavior to a node.
+
+For example:
+
+![](/images/ytt/yaml-primer/map-item-with-value.jpg)
+<!-- source: https://miro.com/app/board/o9J_lIfcKZY=/?moveToWidget=3074457358161105960&cot=14 -->
+_Figure 9: A YAML file with a `ytt` annotation._
+
+Can be understood to mean that the value of `foo` is the result of evaluating the expression `13 + 23 + 6`.
+
+The diagram reveals how the annotation is attached to the YAML tree:
+
+![](/images/ytt/yaml-primer/map-item-ann-with-value-ast.jpg)
+<!-- source: https://miro.com/app/board/o9J_lIfcKZY=/?moveToWidget=3074457358161105960&cot=14 -->
+_Figure 10: The annotation (dotted black) is attached to the **map item** (solid green)._
+
+This is a document, whose value is a map, that contains a single map item. To the right of the map item is an annotation that contains an expression that, when evaluated, becomes a _value_. The annotation is attached to the map item.
+
+### How an Annotation Finds its Node
+
+When attaching annotations, `ytt` follows these two rules (in order):
+1. the annotation is attached to the value-holding node **on its left**, if there is one; otherwise,
```

Comment:
should we spell out what are `value-holding` nodes?

Reply:
We do, 5 lines down.

When I inlined the definition, it got wordy and convoluted: it was both describing the "attachment" process _and_ defining terms.

Perhaps hoist the definition of "value-holding" to the top of this section? 🤔   I'll try that.